File: test/char_code_test.dart
import 'package:comtran/comtran.dart';
import 'package:test/test.dart';

int _punches(String cardCode) {
  final int? p = punchesFromCardCode(cardCode);
  if (p == null) {
    throw ArgumentError.value(cardCode, 'cardCode');
  }
  return p;
}

void main() {
  group('verification anchor (deck-format.md §4.2)', () {
    test(
      'code order reproduces the native collating sequence of J 02.06.16',
      () {
        // The native display's 51 characters, lowest code to highest, from the
        // scan-resolved reading in definition §1.1. Machine specials appear as
        // their names; they have no Set H glyph.
        final List<String> expected = [
          ...'0123456789'.split(''),
          '=', "'", '+',
          ...'ABCDEFGHI'.split(''),
          'plus zero', '.', ')', '-',
          ...'JKLMNOPQR'.split(''),
          'minus zero', r'$', '*', ' ', '/',
          ...'STUVWXYZ'.split(''),
          'record mark', ',', '(', //
        ];
        final actual = <String>[];
        for (var bcd = 0; bcd < 64; bcd++) {
          if (bcd == bcdGroupMark) {
            // Listed only in the Commercial sequence, not the native display.
            continue;
          }
          final String? label = glyphFromBcd(bcd) ?? machineSpecialName(bcd);
          if (label != null) {
            actual.add(label);
          }
        }
        expect(actual, expected);
      },
    );

    test('exactly 12 codes are unattested', () {
      final unattested = <int>[];
      for (var bcd = 0; bcd < 64; bcd++) {
        if (glyphFromBcd(bcd) == null && machineSpecialName(bcd) == null) {
          unattested.add(bcd);
        }
      }
      expect(unattested, [
        0x0A,
        0x0D,
        0x0E,
        0x0F,
        0x1D,
        0x1E,
        0x2D,
        0x2E,
        0x2F,
        0x3D,
        0x3E,
        0x3F,
      ]);
    });
  });

  group('card codes of the source set', () {
    test('the twelve Set H specials read per F p. 12', () {
      const specials = {
        // Card code -> glyph. Blank is the no-punch column, tested below.
        '12': '+',
        '11': '-',
        '11-4-8': '*',
        '0-1': '/',
        '0-4-8': '(',
        '12-4-8': ')',
        '0-3-8': ',',
        '12-3-8': '.',
        '11-3-8': r'$',
        '3-8': '=',
        '4-8': "'",
      };
      specials.forEach((code, glyph) {
        final int? bcd = bcdFromPunches(_punches(code));
        expect(bcd, isNotNull, reason: code);
        expect(glyphFromBcd(bcd!), glyph, reason: code);
        expect(punchesFromBcd(bcd), _punches(code), reason: code);
      });
    });

    test('a blank column is the blank character', () {
      expect(bcdFromPunches(0), bcdBlank);
      expect(glyphFromBcd(bcdBlank), ' ');
      expect(punchesFromBcd(bcdBlank), 0);
    });

    test('letters and digits', () {
      expect(bcdFromPunches(_punches('12-1')), bcdFromGlyph('A'));
      expect(bcdFromPunches(_punches('12-9')), bcdFromGlyph('I'));
      expect(bcdFromPunches(_punches('11-1')), bcdFromGlyph('J'));
      expect(bcdFromPunches(_punches('11-9')), bcdFromGlyph('R'));
      expect(bcdFromPunches(_punches('0-2')), bcdFromGlyph('S'));
      expect(bcdFromPunches(_punches('0-9')), bcdFromGlyph('Z'));
      expect(bcdFromPunches(_punches('0')), bcdFromGlyph('0'));
      expect(bcdFromPunches(_punches('9')), bcdFromGlyph('9'));
      expect(bcdFromGlyph('0'), 0x00);
      expect(punchesFromBcd(0x00), rowBit0);
    });
  });

  group('machine specials (deck-format.md §4.3)', () {
    test('attested card codes', () {
      expect(bcdFromPunches(_punches('12-0')), 0x1A); // plus zero
      expect(bcdFromPunches(_punches('11-0')), 0x2A); // minus zero
      expect(bcdFromPunches(_punches('0-2-8')), 0x3A); // record mark
      expect(bcdFromPunches(_punches('12-5-8')), bcdGroupMark);
      expect(punchesFromBcd(0x1A), _punches('12-0'));
      expect(punchesFromBcd(0x2A), _punches('11-0'));
      expect(punchesFromBcd(0x3A), _punches('0-2-8'));
      expect(punchesFromBcd(bcdGroupMark), _punches('12-5-8'));
    });

    test('specials have no Set H glyph, so no glyph column', () {
      for (final code in ['12-0', '11-0', '0-2-8', '12-5-8']) {
        expect(isGlyphColumn(_punches(code)), isFalse, reason: code);
      }
      expect(isGlyphColumn(_punches('12-1')), isTrue);
      expect(isGlyphColumn(0), isTrue);
    });
  });

  group('read-rule edges (deck-format.md §4.1)', () {
    test('12-8-2 reads as plus zero; the canonical punch stays 12-0', () {
      expect(bcdFromPunches(_punches('12-2-8')), 0x1A);
      expect(punchesFromBcd(0x1A), _punches('12-0'));
    });

    test('12-7-8 has no readout; 7-8 elsewhere keeps the arithmetic rule', () {
      expect(bcdFromPunches(_punches('12-7-8')), isNull);
      expect(bcdFromPunches(_punches('7-8')), 0x0F);
      expect(bcdFromPunches(_punches('11-7-8')), 0x2F);
      expect(bcdFromPunches(_punches('0-7-8')), 0x3F);
    });

    test('octal 35 has no card code', () {
      expect(punchesFromBcd(0x1D), isNull);
    });

    test('illegal combinations have no readout', () {
      expect(bcdFromPunches(_punches('12-11')), isNull);
      expect(bcdFromPunches(rowBitDigit(1) | rowBitDigit(2)), isNull);
      expect(bcdFromPunches(rowBitDigit(8) | rowBitDigit(1)), isNull);
      expect(bcdFromPunches(0xFFF), isNull);
      expect(
        bcdFromPunches(rowBit12 | rowBit0 | rowBitDigit(1)),
        isNull, // 12-0-1: two digit punches under a zone.
      );
    });

    test('every code with a card code round-trips', () {
      for (var bcd = 0; bcd < 64; bcd++) {
        final int? punches = punchesFromBcd(bcd);
        if (punches != null) {
          expect(
            bcdFromPunches(punches),
            bcd,
            reason: 'octal ${bcd.toRadixString(8)}',
          );
        }
      }
    });
  });

  group('card-code strings', () {
    test('rendering uses top-to-bottom row order', () {
      expect(cardCodeFromPunches(_punches('12-5-8')), '12-5-8');
      expect(cardCodeFromPunches(rowBit12 | rowBit11), '12-11');
      expect(cardCodeFromPunches(0), '');
    });

    test('parsing is strict about order and content', () {
      expect(punchesFromCardCode('8-12'), isNull);
      expect(punchesFromCardCode('12-12'), isNull);
      expect(punchesFromCardCode('10'), isNull);
      expect(punchesFromCardCode(''), isNull);
      expect(punchesFromCardCode('12-0'), rowBit12 | rowBit0);
    });
  });
}

File: tools/vscode-punchcard/test/charCode.test.js
'use strict';

const assert = require('node:assert/strict');
const test = require('node:test');

const {
  BCD_BLANK,
  BCD_GROUP_MARK,
  bcdFromGlyph,
  bcdFromPunches,
  cardCodeFromPunches,
  glyphFromBcd,
  isGlyphColumn,
  machineSpecialName,
  punchesFromBcd,
  punchesFromCardCode,
} = require('../out/charCode.js');

/** The 48 source-set characters with their canonical card codes (spec §4.3). */
const SOURCE_SET = [
  ['0', '0'], ['1', '1'], ['2', '2'], ['3', '3'], ['4', '4'],
  ['5', '5'], ['6', '6'], ['7', '7'], ['8', '8'], ['9', '9'],
  ['=', '3-8'], ["'", '4-8'],
  ['+', '12'],
  ['A', '12-1'], ['B', '12-2'], ['C', '12-3'], ['D', '12-4'], ['E', '12-5'],
  ['F', '12-6'], ['G', '12-7'], ['H', '12-8'], ['I', '12-9'],
  ['.', '12-3-8'], [')', '12-4-8'],
  ['-', '11'],
  ['J', '11-1'], ['K', '11-2'], ['L', '11-3'], ['M', '11-4'], ['N', '11-5'],
  ['O', '11-6'], ['P', '11-7'], ['Q', '11-8'], ['R', '11-9'],
  ['$', '11-3-8'], ['*', '11-4-8'],
  [' ', ''],
  ['/', '0-1'],
  ['S', '0-2'], ['T', '0-3'], ['U', '0-4'], ['V', '0-5'], ['W', '0-6'],
  ['X', '0-7'], ['Y', '0-8'], ['Z', '0-9'],
  [',', '0-3-8'], ['(', '0-4-8'],
];

/** The four machine specials with their card codes (spec §4.3). */
const SPECIALS = [
  ['plus zero', '12-0', 0o32],
  ['group mark', '12-5-8', 0o37],
  ['minus zero', '11-0', 0o52],
  ['record mark', '0-2-8', 0o72],
];

function punches(code) {
  return code === '' ? 0 : punchesFromCardCode(code);
}

test('the source set has 48 characters', () => {
  assert.equal(SOURCE_SET.length, 48);
});

test('each source character reads from its card code', () => {
  for (const [glyph, code] of SOURCE_SET) {
    const p = punches(code);
    const bcd = bcdFromPunches(p);
    assert.notEqual(bcd, null, `${code} has no read-out`);
    assert.equal(glyphFromBcd(bcd), glyph, `${code} does not read '${glyph}'`);
  }
});

test('each source character punches its card code', () => {
  for (const [glyph, code] of SOURCE_SET) {
    const bcd = bcdFromGlyph(glyph);
    assert.notEqual(bcd, null, `'${glyph}' has no code`);
    assert.equal(cardCodeFromPunches(punchesFromBcd(bcd)), code);
    assert.equal(isGlyphColumn(punchesFromBcd(bcd)), true);
  }
});

test('a blank column reads octal 60', () => {
  assert.equal(bcdFromPunches(0), BCD_BLANK);
  assert.equal(BCD_BLANK, 0o60);
  assert.equal(glyphFromBcd(BCD_BLANK), ' ');
  assert.equal(punchesFromBcd(BCD_BLANK), 0);
});

test('a bare 0 punch is the digit zero, not a zone', () => {
  assert.equal(bcdFromPunches(punches('0')), 0o00);
  assert.equal(glyphFromBcd(0o00), '0');
});

test('row 0 under a zone is a digit of value ten', () => {
  assert.equal(bcdFromPunches(punches('12-0')), 0o32);
  assert.equal(bcdFromPunches(punches('11-0')), 0o52);
});

test('row 0 above a digit part is the zone', () => {
  assert.equal(bcdFromPunches(punches('0-1')), 0o61);
  assert.equal(bcdFromPunches(punches('0-9')), 0o71);
  assert.equal(bcdFromPunches(punches('0-4-8')), 0o74);
});

test('12-5-8 is the group mark, by the 705 translation', () => {
  assert.equal(bcdFromPunches(punches('12-5-8')), BCD_GROUP_MARK);
  assert.equal(BCD_GROUP_MARK, 0o37);
  assert.equal(glyphFromBcd(BCD_GROUP_MARK), null);
  assert.equal(machineSpecialName(BCD_GROUP_MARK), 'group mark');
});

test('12-7-8 has no read-out', () => {
  assert.equal(bcdFromPunches(punches('12-7-8')), null);
});

test('octal 35 has no card code', () => {
  assert.equal(punchesFromBcd(0o35), null);
});

test('the machine specials keep their names and codes', () => {
  for (const [name, code, octal] of SPECIALS) {
    assert.equal(bcdFromPunches(punches(code)), octal, code);
    assert.equal(machineSpecialName(octal), name);
    assert.equal(glyphFromBcd(octal), null);
  }
});

test('two zone punches have no read-out', () => {
  assert.equal(bcdFromPunches(punches('12-11')), null);
  assert.equal(bcdFromPunches(punches('12-11-0-1-2-3-4-5-6-7-8-9')), null);
});

test('two digit punches without row 8 have no read-out', () => {
  assert.equal(bcdFromPunches(punches('1-2')), null);
  assert.equal(bcdFromPunches(punches('11-3-4')), null);
});

test('the 8-combinations read as the sum of the two punches', () => {
  for (let d = 2; d <= 7; d++) {
    assert.equal(bcdFromPunches(punches(`${d}-8`)), 8 + d);
  }
});

test('every code round-trips through its canonical punches', () => {
  for (let bcd = 0; bcd <= 0o77; bcd++) {
    const p = punchesFromBcd(bcd);
    if (bcd === 0o35) {
      assert.equal(p, null);
      continue;
    }
    assert.equal(bcdFromPunches(p), bcd, `octal ${bcd.toString(8)}`);
  }
});

test('every readable pattern reads a code in range', () => {
  let readable = 0;
  for (let p = 0; p <= 0xfff; p++) {
    const bcd = bcdFromPunches(p);
    if (bcd === null) {
      continue;
    }
    readable++;
    assert.ok(bcd >= 0 && bcd <= 0o77, `pattern ${p} read ${bcd}`);
  }
  // 63 canonical patterns plus the two collisions 12-2-8 and 11-2-8.
  assert.equal(readable, 65);
});

test('card codes are written and parsed top to bottom', () => {
  assert.equal(cardCodeFromPunches(0), '');
  assert.equal(cardCodeFromPunches(0xfff), '12-11-0-1-2-3-4-5-6-7-8-9');
  assert.equal(punchesFromCardCode('12-11-0-1-2-3-4-5-6-7-8-9'), 0xfff);
  assert.equal(punchesFromCardCode('8-12'), null);
  assert.equal(punchesFromCardCode('12-12'), null);
  assert.equal(punchesFromCardCode('13'), null);
  assert.equal(punchesFromCardCode(''), null);
});

test('the two collisions share a code with the zero-punch form', () => {
  assert.equal(bcdFromPunches(punches('12-2-8')), bcdFromPunches(punches('12-0')));
  assert.equal(bcdFromPunches(punches('11-2-8')), bcdFromPunches(punches('11-0')));
  assert.equal(isGlyphColumn(punches('12-2-8')), false);
});
